File: Python/0036_valid_sudoku/0036_valid_sudoku.py
class Solution:
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # check row and col
        for i in range(len(board)):
            row = [False] * 10
            col = [False] * 10
            for j in range(len(board)):
                if board[i][j] != '.':
                    if row[ord(board[i][j]) - ord('0')]:
                        return False
                    row[ord(board[i][j]) - ord('0')] = True         
                if board[j][i] != '.':
                    if (col[ord(board[j][i]) - ord('0')]):
                        return False
                    col[ord(board[j][i]) - ord('0')] = True
        # check 3 x 3 block
        for i in range(3):
            for j in range(3):
                num = [False] * 10
                for row in range(i * 3, i * 3 + 3):
                    for col in range(j * 3, j * 3 + 3):
                        if board[row][col] != '.':
                            if (num[ord(board[row][col]) - ord('0')]):
                                return False
                            num[ord(board[row][col]) - ord('0')] = True
        return True

File: C/0036_valid_sudoku/0036_valid_sudoku.c
bool isValidSudoku(char** board, int boardSize, int* boardColSize)
{
    int *row = malloc(sizeof(int) * 10);
    int *col = malloc(sizeof(int) * 10);
    int *num = malloc(sizeof(int) * 10);
    for (int i = 0; i < boardSize; i++) {
        memset(row, 0, 10 * sizeof(int));
        memset(col, 0, 10 * sizeof(int));
        // check row and col
        for (int j = 0; j < boardSize; j++) {
            if (board[i][j] != '.') {
                if (row[(board[i][j] - '0')])
                    return false;
                row[(board[i][j] - '0')] = true;
            } 
            if (board[j][i] != '.') {
                if (col[(board[j][i] - '0')])
                    return false;
                col[(board[j][i] - '0')] = true;
            }
        }
        // check 3 x 3 block
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                memset(num, 0, sizeof(int) * 10);
                for (int m = 3 * i; m < 3 * i + 3; m++) {
                    for (int n = 3 * j; n < 3 * j + 3; n++) {
                        if (board[m][n] != '.') {
                            if (num[board[m][n] - '0'])
                                return false;
                            num[board[m][n] - '0'] = true;
                        }
                    }
                }
            }
        }    
    }
    free(row);
    free(col);
    free(num);
    return true;
}